src/chainstate/burn/db/burndb.rs

```diff
@@ -0,0 +1,1767 @@
+            }
+        }
+    }
+
+    /// Get all block commitments registered in a block on the burn chain's canonical history.
+    pub fn get_block_commits_by_block(conn: &Connection, block_height: u64) -> Result<Vec<LeaderBlockCommitOp<A, K>>, db_error>
+    where
+        LeaderBlockCommitOp<A, K>: FromRow<LeaderBlockCommitOp<A, K>> + RowOrder
+    {
+        if block_height > ((1 as u64) << 63) - 1 {
+            return Err(db_error::TypeError);
+        }
+
+        let row_order_list : Vec<String> = LeaderBlockCommitOp::row_order().iter().map(|r| format!("block_commits.{}", r)).collect();
+        let row_order = row_order_list.join(",");
+
+        let qry = format!("SELECT {} FROM block_commits JOIN history ON block_commits.txid = history.txid WHERE history.canonical = 1 AND block_commits.block_height = ?1 ORDER BY block_commits.vtxindex ASC", row_order);
+        let args = [&(block_height as i64) as &ToSql];
+
+        BurnDB::<A, K>::query_rows::<LeaderBlockCommitOp<A, K>, _>(conn, &qry.to_string(), &args)
+    }
+    
+    /// Get all user burns registered in a block on the burn chain's canonical history
+    pub fn get_user_burns_by_block(conn: &Connection, block_height: u64) -> Result<Vec<UserBurnSupportOp<A, K>>, db_error>
+    where
+        UserBurnSupportOp<A, K>: FromRow<UserBurnSupportOp<A, K>> + RowOrder
+    {
+        if block_height > ((1 as u64) << 63) - 1 {
+            return Err(db_error::TypeError);
+        }
+
+        let row_order_list : Vec<String> = UserBurnSupportOp::row_order().iter().map(|r| format!("user_burn_support.{}", r)).collect();
+        let row_order = row_order_list.join(",");
+
+        let qry = format!("SELECT {} FROM user_burn_support JOIN history ON user_burn_support.txid = history.txid WHERE history.canonical = 1 AND user_burn_support.block_height = ?1 ORDER BY user_burn_support.vtxindex ASC", row_order);
+        let args = [&(block_height as i64) as &ToSql];
+
+        BurnDB::<A, K>::query_rows::<UserBurnSupportOp<A, K>, _>(conn, &qry.to_string(), &args)
+    }
+
+    /// Find out whether or not a particular VRF key was used before in the canonical burnchain history
+    pub fn has_VRF_public_key(conn: &Connection, key: &VRFPublicKey) -> Result<bool, db_error> {
+        let qry = "SELECT COUNT(leader_keys.public_key) FROM leader_keys JOIN history ON leader_keys.txid = history.txid WHERE history.canonical = 1 AND leader_keys.public_key = ?1 AND history.canonical = 1";
+        let args = [&ECVRF_public_key_to_hex(&key)];
+        let count = BurnDB::<A, K>::query_count(conn, &qry.to_string(), &args)?;
+        Ok(count != 0)
+    }
+
+    /// Find out whether or not a given consensus hash is "recent" enough to be used in the
+    /// canonical burnchain history
+    pub fn is_fresh_consensus_hash(conn: &Connection, current_block_height: u64, consensus_hash: &ConsensusHash) -> Result<bool, db_error> {
```

Comment:
There's a couple functions in this object that would probably be more clear if they could be separated from the SQL queries somewhat. And not just for the sake of code clarity-- trying to interact with the database implementation more abstractly will be helpful if we ever want to use a different database implementation in the future.

`is_fresh_consensus_hash` -> it seems like the logic that decides whether or not a consensus hash is fresh enough shouldn't be done at the database level (this one has led to bugs before) -- as in, the database should expose a `get_consensus_hash_blockheight` function, and something else should be checking it.

`get_prev_consensus_hash / make_consensus_hash` -> this also seems like its pretty easy to separate from the database.

To me, it seems like the database object is just doing too much -- it's a SQL wrapper, but it's _also_ implementing some consensus logic, and responsible for checking transactions and processing Bitcoin blocks.

Reply:
Sure -- I can move these to the `ConsensusHash` implementation.

Reply:
I think the transaction checking and processing parts should likely move as well:

```
process_block
check_transaction
```

And probably these:
```
parse_transaction
commit_transaction
```

Commit transaction is kind of confusing -- because there's another notion in here of a transaction that could be committed (the database transaction started in `begin_tx`, I think).

Reply:
Sounds good.  I'll create a `Burnchain` type that has an associated burn database and contains these high-level operations.

`commit_transaction` just reuses nomenclature from the previous implementation.  In reality, all this code does is insert the transaction in the context of a SQL transaction.